nncf/experimental/tensor/functions.py

```diff
@@ -324,28 +326,113 @@ def zeros_like(a: TTensor) -> TTensor:
     return Tensor(zeros_like(a.data))
 
 
-__all__ = [
-    "device",
-    "squeeze",
-    "flatten",
-    "max",
-    "min",
-    "abs",
-    "astype",
-    "reshape",
-    "all",
-    "allclose",
-    "any",
-    "count_nonzero",
-    "isempty",
-    "isclose",
-    "maximum",
-    "minimum",
-    "ones_like",
-    "minimum",
-    "where",
-    "zeros_like",
-]
+@functools.singledispatch
+def stack(x: List[Tensor], axis: int = 0) -> Tensor:
+    """
+    Stacks a list of Tensors rank-R tensors into one Tensor rank-(R+1) tensor.
+
+    :param x: List of Tensors.
+    :param axis: The axis to stack along.
+    :return: Stacked Tensor.
+    """
+    if isinstance(x, List):
+        unwrapped_x = [i.data for i in x]
```

Comment:
Can this go into `_dispatch_list` too?

Reply:
`_dispatch_list` have one task to run `dispatch` by type of element from first argument that has type List. 

Why `_dispatch_list` should contain extra logic to unwrap data? 
Should it unwrap only first argument or all? What about kwargs?  
I think unwrapping data should be done where we know names to avoid call `isinstance` check for all arguments of function.

Reply:
Limit the `_dispatch_list` to taking lists of tensors as its first arg. You don't need to build the most generic API here.

Reply:
`_dispatch_list` already limited to get function and args to dispatch by first element in the first argument.

Reply:
It is an implicit limitation. Make it explicit by separating the corresponding argument from *args in the signature and setting up a corresponding type hint.

Reply:
Please write the justification to your suggestion.

Reply:
That would be code reuse and explicitness through typing.

Reply:
It is not a reason to unwrap data in `_dispatch_list` function.
Is it blocking the merge?

Reply:
Yes, it is a reason, since you have to take every opportunity at making the code better, especially when you are pointed toward such opportunity. Now that you have asked - yes, I want to see this change in this PR, unless *you* provide a reason as to why your solution is better than what I suggest.

Reply:
I am requesting the changes in accordance with the arguments I've provided in [two](https://github.com/openvinotoolkit/nncf/pull/2058#discussion_r1345596652) [comments](https://github.com/openvinotoolkit/nncf/pull/2058#discussion_r1345882442) now.

Reply:
@AlexKoff88 Need you help to resolve it. Does this comment block the merge of this PR or not?
@alexsu52

Reply:
@vshampor, please provide a code of your proposal.

Reply:
In the `_dispatch_list` definition, you would have:
```python
def _dispatch_list(fn: "functools._SingleDispatchCallable", tensor_list: List[Tensor], *args, **kwargs):
    unwrapped_list = [i.data for i in tensor_list]
    return fn.dispatch(type(unwrapped_list[0]))(unwrapped_list, *args, **kwargs)
```
and the `stack` implementation then becomes simply:

```python
@functools.singledispatch
def stack(x: List[Tensor], axis: int = 0) -> Tensor:
    return Tensor(_dispatch_list(stack, x, axis=axis))
```
and this is exactly what I wrote [here](https://github.com/openvinotoolkit/nncf/pull/2058#discussion_r1343002474) and [here](https://github.com/openvinotoolkit/nncf/pull/2058#discussion_r1345467003), in good English. I would expect at this point in time of the NNCF devs to be able to do this kind of primitive, obvious refactoring on their own without code examples and instead to use their reading comprehension skills. As it is, note that I had to spend (a quite marginal) amount of time to write this code on my own to demonstrate this.

Reply:
ok, added